### PR TITLE
Added pca transformations.

### PR DESCRIPTION
**Pull request by latture**
This PR adds 2 PCA data transformations: one from sklearn and one from prince. I added prince for 2 reasons: 1) it has some configuration parameters that rescale data automatically (makes it a bit easier to use when compared to sklearn), and 2) prince contains a multiple-factor analysis implementation that I'd like to investigate and potentially add in a subsequent PR. I also made both PCA implementations invertible, so we can recover data in the original dimensionality.

### CODE REVIEW

File: smlb/decomposition/pca_prince.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    """Principal component analysis (PCA), prince implementation.
+
+    .. seealso::
+        See `here <https://github.com/MaxHalford/prince#principal-component-analysis-pca>`_ for full documentation.
+    """
+
+    def __init__(
+        self,
+        rng: Optional[int] = None,
+        rescale_with_mean: bool = True,
```
Comment (gregor-robinson) [inline smlb/decomposition/pca_prince.py:25]: "Rescale" sounds to me like an invertible linear transformation rather than a translation. How about calling this `subtract_mean`?

File: smlb/decomposition/pca_prince.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        rescale_with_std: bool = True,
```
Comment (gregor-robinson) [inline smlb/decomposition/pca_prince.py:26]: How about calling this `normalize_std`?

File: smlb/decomposition/pca_prince.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        """
+
+        super().__init__(*args, **kwargs)
+        rescale_with_mean = params.boolean(rescale_with_mean)
+        rescale_with_std = params.boolean(rescale_with_std)
+        n_components = params.integer(n_components, from_=1)
+        n_iter = params.integer(n_iter, from_=1)
+        inverse_transform = params.boolean(inverse_transform)
+
+        self._inverse_transform: bool = inverse_transform
```
Comment (bfolie) [inline smlb/decomposition/pca_prince.py:51]: collapse into one line?

File: smlb/decomposition/pca_sklearn.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            data: data to fit
+
+        Returns:
+            the instance itself
+        """
+        data = params.instance(data, Data)
+        n = data.num_samples
+
+        xtrain = params.real_matrix(data.samples(), nrows=n)
+        ytrain = params.real_vector(data.labels(), dimensions=n)
```
Comment (bfolie) [inline smlb/decomposition/pca_sklearn.py:63]: If the data have multiple labels, would this result in an error because it can't be cast to a vector?

File: smlb/decomposition/pca_prince.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        Parameters:
+            data: data to fit
+
+        Returns:
+            the instance itself
+        """
+        data = params.instance(data, Data)
+
+        xtrain = params.real_matrix(data.samples(), nrows=data.num_samples)
+        ytrain = params.real_vector(data.labels(), dimensions=data.num_samples)
```
Comment (bfolie) [inline smlb/decomposition/pca_prince.py:73]: If the data have multiple labels, would this result in an error because it can't be cast to a vector?

File: smlb/decomposition/pca_prince.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        **kwargs
+    ):
+        """Initialize state.
+
+        Parameters:
+            rng: Random state
+            rescale_with_mean: Whether to subtract each column's mean or not.
+            rescale_with_std: Whether to divide each column by it's standard deviation or not.
+            n_components: The number of principal components to compute.
+            n_iter: The number of iterations used for computing the SVD.
```
Comment (gregor-robinson) [inline smlb/decomposition/pca_prince.py:40]:
```suggestion
            n_iter: The number of power iterations used for computing the SVD.
```

So few iterations are usually not sufficient for typical applications of the power method, so I was curious whether this might be enough. Tl;dr: probably. See page 10 of https://arxiv.org/pdf/0909.4061.pdf

File: smlb/decomposition/pca_prince.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        inverse_transform: bool = False,
+        *args,
+        **kwargs
+    ):
+        """Initialize state.
+
+        Parameters:
+            rng: Random state
+            rescale_with_mean: Whether to subtract each column's mean or not.
+            rescale_with_std: Whether to divide each column by it's standard deviation or not.
```
Comment (gregor-robinson) [inline smlb/decomposition/pca_prince.py:38]:
```suggestion
            rescale_with_std: Whether to divide each column by its standard deviation or not.
```

File: smlb/decomposition/pca_sklearn.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+        Parameters:
+            rng: Random state
+            n_components: Number of components to keep.
+                          If ``None``, ``n_components == min(n_samples, n_features) - 1`` when fit.
+            inverse_transform: Whether to perform inverse transformation. Default is ``False``
+        """
+
+        super().__init__(*args, **kwargs)
+        n_components = params.optional_(n_components, self.nneg_int_f)
```
Comment (bfolie) [inline smlb/decomposition/pca_sklearn.py:44]: This seems to allow n_components=0. What does it mean to have 0 components?

File: smlb/decomposition/pca_sklearn.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
+from copy import deepcopy
+from typing import Optional
+
+from sklearn import decomposition
```
Comment (bfolie) [inline smlb/decomposition/pca_sklearn.py:4]: why not import `decomposition.PCA` directly?

File: smlb/decomposition/pca_prince.py (hunk 6 above)
Comment (bfolie) [inline smlb/decomposition/pca_prince.py:40]: Any guidance on how many to use? Why is the default 3?

File: smlb/decomposition/pca_sklearn.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                          If ``None``, ``n_components == min(n_samples, n_features) - 1`` when fit.
+            inverse_transform: Whether to perform inverse transformation. Default is ``False``
+        """
+
+        super().__init__(*args, **kwargs)
+        n_components = params.optional_(n_components, self.nneg_int_f)
+        inverse_transform = params.boolean(inverse_transform)
+
+        self._inverse_transform: bool = inverse_transform
+        self._pca: decomposition.PCA = decomposition.PCA(n_components, random_state=rng)
```
Comment (gregor-robinson) [inline smlb/decomposition/pca_sklearn.py:48]: Should this also have standardization options? Power iteration control?

File: smlb/decomposition/pca_sklearn.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        *args,
+        **kwargs
+    ):
+        """Create a PCA instance.
+
+        Parameters:
+            rng: Random state
+            n_components: Number of components to keep.
+                          If ``None``, ``n_components == min(n_samples, n_features) - 1`` when fit.
+            inverse_transform: Whether to perform inverse transformation. Default is ``False``
```
Comment (gregor-robinson) [inline smlb/decomposition/pca_sklearn.py:40]: Like above, I don't think this belongs in the constructor as designed.

File: smlb/decomposition/pca_sklearn.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        inverse_transform = params.boolean(inverse_transform)
+
+        self._inverse_transform: bool = inverse_transform
+        self._pca: decomposition.PCA = decomposition.PCA(n_components, random_state=rng)
+
+    def fit(self, data: Data) -> "PCASklearn":
+        """Fit the model with input ``data``.
+
+        Parameters:
+            data: data to fit
```
Comment (gregor-robinson) [inline smlb/decomposition/pca_sklearn.py:54]:
```suggestion
            data: data on which to compute PCA
```

File: smlb/decomposition/pca_prince.py (hunk 6 above)
Comment (gregor-robinson) [reply smlb/decomposition/pca_prince.py:40]: So few iterations are usually not sufficient for typical applications of the power method, so I was curious whether this might be enough. Tl;dr: probably. See page 10 of https://arxiv.org/pdf/0909.4061.pdf

File: smlb/decomposition/pca_prince.py (hunk 1 above)
Comment (latture) [reply smlb/decomposition/pca_prince.py:25]: I use `rescale_with_mean` and `rescale_with_std` since those are the parameter names prince uses. This is just a passthrough class to their implementation, so I think renaming their parameters would make it more difficult to use.

File: smlb/decomposition/pca_sklearn.py (hunk 8 above)
Comment (latture) [reply smlb/decomposition/pca_sklearn.py:44]: That's an error, it should be >0. I'll update the code.

File: smlb/decomposition/pca_sklearn.py (hunk 9 above)
Comment (latture) [reply smlb/decomposition/pca_sklearn.py:4]: I originally called this class `PCA` before I added the second implementation. I'll import it directly.

File: smlb/decomposition/pca_prince.py (hunk 6 above)
Comment (latture) [reply smlb/decomposition/pca_prince.py:40]: 3 is the default used by the parent project, so I carried that over as the default here. I have not looked into what the optimal amount is.

File: smlb/decomposition/pca_sklearn.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+    @staticmethod
+    def nneg_int_f(arg):
+        return params.integer(arg, from_=0)
+
+    def __init__(
+        self,
+        rng: Optional[int] = None,
+        n_components: Optional[int] = None,
+        inverse_transform: bool = False,
```
[the remark this answers is not in the document]
Comment (latture) [reply smlb/decomposition/pca_sklearn.py:31]: I think that would be a cleaner implementation. I'll update the constructor of this and the other PCA class.

File: smlb/decomposition/pca_sklearn.py (hunk 10 above)
Comment (latture) [reply smlb/decomposition/pca_sklearn.py]: Sure, I can expose the other parameters of sklearn's implementation.

**Review by gregor-robinson — APPROVED** [review]

**Review by bfolie — APPROVED** [review]

File: smlb/decomposition/pca_prince.py (hunk 5 above)
Comment (latture) [reply smlb/decomposition/pca_prince.py:70]: Yes, it would cause an error. After looking through the `.fit` documentation, it appears `y` is ignored, so we only need to pull out `x` from the data. I'll push the change.

**Review by bfolie — APPROVED** [review]
